src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+use crate::state;
+use crate::state::{ArchiveData, ArchiveState, DeviceDataInternal};
+use candid::Principal;
+use ic_cdk::api::call::CallResult;
+use ic_cdk::api::management_canister::main::{
+    canister_status, create_canister, install_code, CanisterIdRecord, CanisterInstallMode,
+    CanisterInstallMode::Install, CreateCanisterArgument, InstallCodeArgument,
+};
+use ic_cdk::api::time;
+use ic_cdk::{id, notify};
+use internet_identity_interface::{
+    ArchiveInit, DeployArchiveResult,
+    DeployArchiveResult::{CreationFailed, UpgradeFailed},
+    DeviceDataUpdate, Entry, Operation, Private, UserNumber,
+};
+use serde_bytes::ByteBuf;
+use sha2::Digest;
+use sha2::Sha256;
+use CanisterInstallMode::Upgrade;
+
+pub async fn deploy_archive(wasm: ByteBuf) -> DeployArchiveResult {
+    let archive_state = state::persistent_state_mut(|persistent_state| {
+        let archive_state = persistent_state.archive_info.state.clone();
+        if archive_state == ArchiveState::NotCreated {
+            // lock archive creation because of async operation
+            persistent_state.archive_info.state = ArchiveState::CreationInProgress;
+        }
+        archive_state
+    });
+
+    match archive_state {
+        ArchiveState::NotCreated => create_and_install_archive(wasm.into_vec()).await,
+        ArchiveState::CreationInProgress => DeployArchiveResult::CreationInProgress,
+        ArchiveState::Created(data) => {
+            match install_archive(data.archive_canister, wasm.into_vec()).await {
+                Ok(()) => DeployArchiveResult::Success(data.archive_canister),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_and_install_archive(wasm: Vec<u8>) -> DeployArchiveResult {
+    match create_archive().await {
+        Err(e) => {
+            state::persistent_state_mut(|persistent_state| {
+                // unlock archive creation again
+                persistent_state.archive_info.state = ArchiveState::NotCreated
+            });
+            return CreationFailed(format!("failed to create archive: {:?}", e));
+        }
+        Ok(data) => {
+            let archive_canister_id = data.archive_canister;
+            state::persistent_state_mut(|persistent_state| {
+                // safe archive info permanently
+                persistent_state.archive_info.state = ArchiveState::Created(data)
+            });
+            match install_archive(archive_canister_id, wasm).await {
+                Ok(()) => DeployArchiveResult::Success(archive_canister_id),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_archive() -> CallResult<ArchiveData> {
+    let (result,) = create_canister(CreateCanisterArgument { settings: None }).await?;
```

Comment:
Might have missed something, but it looks like we're creating the canister before even checking the hash. Does that mean that an II canister with no hash specified will still be creating an archive canister if anyone hits the `deploy_archive` endpoint?

Reply:
Yes, that is true. I will change that.